fragment.go

```diff
@@ -1001,32 +1129,54 @@ func (f *fragment) rangeLT(bitDepth uint, predicate uint64, allowEquality bool)
 			if bit == 0 {
 				return keep, nil
 			}
-			return b.Difference(row.Difference(keep)), nil
+			return filter.Difference(row.Difference(keep)), nil
 		}
 
 		// If bit is zero then remove all set columns not in excluded bitmap.
 		if bit == 0 {
-			b = b.Difference(row.Difference(keep))
+			filter = filter.Difference(row.Difference(keep))
 			continue
 		}
 
 		// If bit is set then add columns for set bits to exclude.
 		// Don't bother to compute this on the final iteration.
 		if i > 0 {
-			keep = keep.Union(b.Difference(row))
+			keep = keep.Union(filter.Difference(row))
 		}
 	}
 
-	return b, nil
+	return filter, nil
 }
 
-func (f *fragment) rangeGT(bitDepth uint, predicate uint64, allowEquality bool) (*Row, error) {
-	b := f.row(uint64(bitDepth))
+func (f *fragment) rangeGT(bitDepth uint, predicate int64, allowEquality bool) (*Row, error) {
+	b := f.row(bsiExistsBit)
+
+	// Create predicate without sign bit.
+	upredicate := uint64(predicate)
+	if predicate < 0 {
+		upredicate = uint64(-predicate)
+	}
+
+	// If predicate is positive, return all positives greater than predicate.
+	if (predicate >= 0 && allowEquality) || (predicate >= -1 && !allowEquality) {
+		return f.rangeGTUnsigned(b.Difference(f.row(bsiSignBit)), bitDepth, upredicate, allowEquality)
+	}
+
+	// If predicate is negative, return all negatives less than than upredicate and all positives.
+	neg, err := f.rangeLTUnsigned(b.Intersect(f.row(bsiSignBit)), bitDepth, upredicate, allowEquality)
+	if err != nil {
+		return nil, err
+	}
+	pos := b.Difference(f.row(bsiSignBit))
```

Comment:
this has me pondering the desireability of an any-negative-values flag for the field, because if there's no negative values, we could save a lot of time here.